File: cli/src/fno/plan/_status.py
"""Frontmatter status state machine for lean single-doc plan architecture.

Enforces the monotonic progression:
    design -> ready -> in_progress -> shipped

Tracks the same lifecycle ladder the graph `_status` speaks, with `shipped` as
the plan-side name for the graph's `in_review` rung (GRAPH_TO_PLAN_STATUS is
the alignment layer). `idea` has no plan doc so it never appears here, and
`done`/`archived` are off-axis terminals.

`reviewing`/`shipping` were pruned (x-f34f): they had zero consumers and the
graph has no derived state that distinguishes them, so they never got written.
The reconcile sweep now folds them into `shipped` as Tier-1 synonyms.

Backward transitions, identity transitions, and unknown statuses all raise
StatusTransitionError. No silent fallbacks.
"""

from __future__ import annotations

from typing import Any, Optional

STATUS_PROGRESSION: tuple[str, ...] = (
    "design",
    "ready",
    "in_progress",
    "shipped",
)

# Off-axis terminals: written directly (graduate stamps `done`; the status
# sweep stamps `archived`), NOT part of the monotonic axis. Inserting either
# into STATUS_PROGRESSION would break the forward-transition index math.
TERMINAL_STATUSES: tuple[str, ...] = ("done", "archived")

# The full canonical plan-status vocabulary: axis + terminals. The reconcile
# sweep leaves any status in this set untouched (it corrects drift only).
KNOWN_STATUSES: frozenset[str] = frozenset(STATUS_PROGRESSION) | frozenset(TERMINAL_STATUSES)

# Graph derived `_status` -> plan `status` projection (x-f34f). Total over the
# graph vocabulary; None means "no plan write" (a graph-side gate that must not
# touch plan state). Since x-5d91 the two vocabularies are the same ladder, so
# this is near-identity - it survives to carry the genuinely non-identity rows
# (idea -> design, superseded -> archived, and the two None gates).
GRAPH_TO_PLAN_STATUS: dict[str, str | None] = {
    "idea": "design",  # node exists, no plan doc yet
    "design": "design",  # doc exists but is still a design doc
    "ready": "ready",
    "in_progress": "in_progress",
    "claimed": "in_progress",  # legacy graph vocabulary, pre-x-5d91 rows
    "blocked": None,  # graph-side gate; plan keeps its current state
    "in_review": "shipped",  # PR open = implementation complete
    "done": "done",  # merged
    "superseded": "archived",
    "deferred": None,  # pause is reversible; plan state stands
}

# Forward-only ordering for the projection. `done` caps the axis; `archived`
# (from `superseded`) is a terminal reachable from any non-terminal state and is
# never rank-compared.
_PROJECTION_RANK: dict[str, int] = {
    "design": 0,
    "ready": 1,
    "in_progress": 2,
    "shipped": 3,
    "done": 4,
}


def _norm_status(raw: object) -> str:
    """Bare lowercase token from a raw frontmatter status value."""
    return str(raw if raw is not None else "").strip().strip("'\"").lower()


def project_plan_status(current: object, graph_status: str) -> Optional[str]:
    """Plan status to WRITE for a node in ``graph_status``, or None to leave it.

    Forward-only along design < ready < in_progress < shipped < done. Returns
    None when the graph status maps to no write, the target equals the current
    status, or the target would be a backward move (graph wins forward, a human
    hand-edit wins backward). ``archived`` is written over any non-terminal
    plan state (superseded) but never over ``done`` or ``archived``.
    """
    target = GRAPH_TO_PLAN_STATUS.get(graph_status)
    if not target:
        return None
    cur = _norm_status(current)
    if target == cur:
        return None
    if target == "archived":
        return None if cur in ("done", "archived") else "archived"
    # target is a forward-axis status (design..done)
    if cur in ("done", "archived"):
        return None  # terminal: never auto-rewritten forward off a terminal
    if _PROJECTION_RANK[target] <= _PROJECTION_RANK.get(cur, -1):
        return None
    return target


class StatusTransitionError(ValueError):
    """Raised on invalid status transitions."""


def validate_transition(old: str, new: str) -> None:
    """Raise StatusTransitionError on:

    - unknown old or new status
    - backward transition (new index < old index)
    - identity transition (old == new)

    Allow: forward transitions (new index > old index) by any number of steps.
    """
    if old not in STATUS_PROGRESSION:
        raise StatusTransitionError(
            f"Unknown status {old!r}. Valid statuses: {list(STATUS_PROGRESSION)}"
        )
    if new not in STATUS_PROGRESSION:
        raise StatusTransitionError(
            f"Unknown status {new!r}. Valid statuses: {list(STATUS_PROGRESSION)}"
        )

    old_index = STATUS_PROGRESSION.index(old)
    new_index = STATUS_PROGRESSION.index(new)

    if new_index == old_index:
        raise StatusTransitionError(
            f"Identity transition rejected: status is already {old!r}. "
            "Provide a different target status."
        )

    if new_index < old_index:
        raise StatusTransitionError(
            f"Backward transition rejected: cannot move from {old!r} (index {old_index}) "
            f"to {new!r} (index {new_index}). "
            f"Status progression is monotonic: {' -> '.join(STATUS_PROGRESSION)}"
        )


def coerce_status_from_yaml(value: Any) -> str:
    """Coerce a raw yaml.safe_load value to a valid status string.

    yaml.safe_load returns Python True for unquoted `status: true` in YAML.
    This function handles that by coercing:
      - bool -> lowercase string ("true" / "false")
      - None -> raises StatusTransitionError
      - anything else -> str(value)

    Then validates the result is in STATUS_PROGRESSION. Raises
    StatusTransitionError if the coerced value is not a known status.

    Per feedback_literal_string_rejects_yaml_bool memory entry.
    """
    if value is None:
        raise StatusTransitionError(
            "Status value is None; expected one of: "
            + ", ".join(repr(s) for s in STATUS_PROGRESSION)
        )

    if isinstance(value, bool):
        coerced = str(value).lower()  # True -> "true", False -> "false"
    else:
        coerced = str(value)

    if coerced not in STATUS_PROGRESSION:
        raise StatusTransitionError(
            f"Unknown status {coerced!r} (coerced from {value!r}). "
            f"Valid statuses: {list(STATUS_PROGRESSION)}"
        )

    return coerced

File: crates/fno-agents/src/plan_doc/status.rs
//! Plan-frontmatter status projection, ported 1:1 from `cli/src/fno/plan/_status.py`.
//!
//! The monotonic plan axis is design < ready < in_progress < in_review < done;
//! `superseded` is an off-axis terminal. The retired spellings resolve on read
//! via the shared alias table in graph_store.

/// Retired plan spellings, accepted on read and never written. Reuses the
/// store's table (the Python reader `plan._status.STATUS_ALIASES` is the same
/// three rows) instead of retyping it.
const STATUS_ALIASES: &[(&str, &str)] = crate::graph_store::PLAN_STATUS_ALIASES;

/// Graph derived `_status` -> plan `status`. None means "no plan write".
const GRAPH_TO_PLAN_STATUS: &[(&str, Option<&str>)] = &[
    ("idea", Some("idea")),
    ("design", Some("design")),
    ("ready", Some("ready")),
    ("in_progress", None),
    ("claimed", None),
    ("blocked", None),
    ("in_review", Some("in_review")),
    ("done", Some("done")),
    ("superseded", Some("superseded")),
    ("deferred", None),
];

/// Forward-only ordering, keyed by the plan vocabulary. `idea` sits at -1, the
/// same rank an unknown/absent status gets, so it is never a projection target.
fn projection_rank(status: &str) -> i32 {
    match status {
        "design" => 0,
        "ready" => 1,
        "in_progress" => 2,
        "in_review" => 3,
        "done" => 4,
        _ => -1,
    }
}

/// Normalized status: bare lowercase token, quotes stripped.
fn norm_status(raw: Option<&str>) -> String {
    raw.unwrap_or("")
        .trim()
        .trim_matches(|c| c == '\'' || c == '"')
        .to_lowercase()
}

/// Normalized status with any retired spelling resolved to its survivor.
/// Read-path translation only: callers compare and rank against the result,
/// they never write it back over the doc that supplied it.
pub fn canonical_status(raw: Option<&str>) -> String {
    let s = norm_status(raw);
    STATUS_ALIASES
        .iter()
        .find(|(from, _)| *from == s)
        .map(|(_, to)| (*to).to_string())
        .unwrap_or(s)
}

/// Plan status to WRITE for a node in `graph_status`, or None to leave it.
/// Forward-only; returns None when the graph status maps to no write, the
/// target equals the current status, or the target would be a backward move.
/// `superseded` is written over any non-terminal plan state but never over
/// `done` or `superseded`.
pub fn project_plan_status(current: Option<&str>, graph_status: &str) -> Option<String> {
    let target = GRAPH_TO_PLAN_STATUS
        .iter()
        .find(|(from, _)| *from == graph_status)
        .and_then(|(_, to)| *to)?;
    let cur = canonical_status(current);
    if target == cur {
        return None;
    }
    if target == "superseded" {
        return if cur == "done" || cur == "superseded" {
            None
        } else {
            Some("superseded".to_string())
        };
    }
    if cur == "done" || cur == "superseded" {
        return None; // terminal: never auto-rewritten forward off a terminal
    }
    if projection_rank(&target) <= projection_rank(&cur) {
        return None;
    }
    Some(target.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn aliases_resolve_on_read() {
        assert_eq!(canonical_status(Some("shipped")), "in_review");
        assert_eq!(canonical_status(Some("archived")), "superseded");
        assert_eq!(canonical_status(Some("stub")), "idea");
        assert_eq!(canonical_status(Some("Ready")), "ready");
        assert_eq!(canonical_status(Some("'done'")), "done");
        assert_eq!(canonical_status(None), "");
        assert_eq!(canonical_status(Some("weird")), "weird");
    }

    #[test]
    fn forward_only_projection() {
        assert_eq!(
            project_plan_status(Some("ready"), "in_review"),
            Some("in_review".into())
        );
        assert_eq!(
            project_plan_status(Some("shipped"), "done"),
            Some("done".into())
        );
        // Backward moves refused.
        assert_eq!(project_plan_status(Some("in_review"), "claimed"), None);
        assert_eq!(project_plan_status(Some("done"), "ready"), None);
        // Gated states write nothing.
        assert_eq!(project_plan_status(Some("ready"), "blocked"), None);
        assert_eq!(project_plan_status(Some("ready"), "deferred"), None);
        assert_eq!(project_plan_status(Some("ready"), "in_progress"), None);
        // Identity is a no-op.
        assert_eq!(project_plan_status(Some("ready"), "ready"), None);
        // Terminals never rewritten forward.
        assert_eq!(project_plan_status(Some("superseded"), "done"), None);
    }

    #[test]
    fn superseded_written_over_non_terminal_only() {
        assert_eq!(
            project_plan_status(Some("ready"), "superseded"),
            Some("superseded".into())
        );
        assert_eq!(project_plan_status(Some("done"), "superseded"), None);
        assert_eq!(project_plan_status(Some("superseded"), "superseded"), None);
    }

    #[test]
    fn idea_is_never_a_target() {
        // rank(idea) = -1: never above any current rung, and unknown == idea rank.
        assert_eq!(project_plan_status(Some("design"), "idea"), None);
        assert_eq!(project_plan_status(None, "idea"), None);
    }
}
